src/mlpack/methods/ann/layer/group_norm_impl.hpp

```diff
@@ -0,0 +1,164 @@
+#define MLPACK_METHODS_ANN_LAYER_GROUPNORM_IMPL_HPP
+
+// In case it is not included.
+#include "group_norm.hpp"
+
+namespace mlpack {
+namespace ann { /** Artificial Neural Network. */
+
+
+template<typename InputDataType, typename OutputDataType>
+GroupNorm<InputDataType, OutputDataType>::GroupNorm() :
+    size(0),
+    groupCount(1),
+    eps(1e-8),
+    loading(false)
+{
+  // Nothing to do here.
+}
+
+template <typename InputDataType, typename OutputDataType>
+GroupNorm<InputDataType, OutputDataType>::GroupNorm(
+    const size_t size, const size_t groupCount, const double eps) :
+    size(size),
+    groupCount(groupCount),
+    eps(eps),
+    loading(false)
+{
+  if (size % groupCount != 0)
+  {
+    Log::Fatal << "Total input Size must be divisible by groupCount!" << std::endl;
+  }
+
+  weights.set_size(size * groupCount * 2 + 1, 1);
+}
+
+template<typename InputDataType, typename OutputDataType>
+void GroupNorm<InputDataType, OutputDataType>::Reset()
+{
+  gamma = arma::mat(weights.memptr(), size * groupCount, 1, false, false);
+  beta = arma::mat(weights.memptr() + gamma.n_elem, size * groupCount, 1, false, false);
+
+  if (!loading)
+  {
+    gamma.fill(1.0);
+    beta.fill(0.0);
+  }
+
+  loading = false;
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void GroupNorm<InputDataType, OutputDataType>::Forward(
+    const arma::Mat<eT>& input, arma::Mat<eT>& output)
+{
+  arma::mat reshapedInput((input).memptr(),
+    size / groupCount, groupCount * input.n_cols, false, false);
+
+  if (output.is_empty())
+    output.zeros(size, input.n_cols);
+
+  arma::mat reshapedOutput((output).memptr(),
+    size / groupCount, groupCount * output.n_cols, false, false);
+
+  mean = arma::mean(reshapedInput, 0);
+  variance = arma::var(reshapedInput, 1, 0);
+
+  // Normalize the input.
+  reshapedOutput = reshapedInput.each_row() - mean;
+  inputMean = reshapedOutput;
+  reshapedOutput.each_row() /= arma::sqrt(variance + eps);
+
+  // Reused in the backward and gradient step.
+  normalized = output;
+
+  // Scale and shift the output.
+  reshapedOutput.each_col() %= gamma;
+  reshapedOutput.each_col() += beta;
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void GroupNorm<InputDataType, OutputDataType>::Backward(
+    const arma::Mat<eT>& input, const arma::Mat<eT>& gy, arma::Mat<eT>& g)
+{
+  arma::mat inputReshaped((input).memptr(),
+    size / groupCount, groupCount * input.n_cols, false, false);
+  arma::mat gyReshaped((gy).memptr(),
+    size / groupCount, groupCount * gy.n_cols, false, false);
+  arma::mat gReshaped((g).memptr(),
+    size / groupCount, groupCount * g.n_cols, false, false);
+
+  const arma::mat stdInv = 1.0 / arma::sqrt(variance + eps);
+
+  // dl / dxhat.
+  const arma::mat norm = gyReshaped.each_col() % gamma;
+
+  // sum dl / dxhat * (x - mu) * -0.5 * stdInv^3.
+  const arma::mat var = arma::sum(norm % inputMean, 0) %
```

Comment:
Don't we have to take the sum over the columns here?

Reply:
yes. If the batch is of size **m** then there will be **m** columns.  Let's say one sample size = **n**, and we want to divide it into **g** groups. Then  the initial input will be a matrix of size **n * m**. Now we can reshape this to
**(n/g) * (m*g)**. Then apply layer norm to this. This is equivalent to group norm.
By reshaping we are dividing one sample size **(n*1)** to **(n/g)*g**.

Reply:
Thanks for the explanation.